language/stdlib/modules/Offer.move

```diff
@@ -35,6 +35,96 @@ module Offer {
     borrow_global<T<Offered>>(offer_address).for
   }
+
+  /**
+   **************** SPECIFICATIONS ****************
+
+  This module defines a resource T that is used as a permissioned trading scheme between accounts.
+  It defines two main functions for creating and retrieving a struct offered by some user
+  inside the resource T under the offerer's account.
+
+  Currently, the only other module that depends on this module is LibraConfig, where it's used to
+  pass a capability to an account that allows it to modify a config.
+  */
+
+  /// # Module specification
+
+  spec module {
+    /// Verify all functions in this module
+    pragma verify = true;
+    /// Helper function that returns whether or not the `recipient` is an intended
+    /// recipient of the offered struct in the T<Offered> resource at the address `offer_address`
+    /// Returns true if the recipient is allowed to redeem T<Offered> at `offer_address`
+    /// and false otherwise.
+    define is_allowed_recipient<Offered>(offer_addr: address, recipient: address): bool {
+      recipient == global<T<Offered>>(offer_addr).for || recipient == offer_addr
+    }
+  }
+
+  /// ## Creation of Offers
+
+  spec schema OnlyCreateCanCreateOffer {
+    /// Only `Self::create` can create an Offer.T under an address.
+    ///
+    /// **Informally:** No function to which this is applied can create an offer.
+    /// If there didn't exist an Offer under some `addr`, then it continues
+    /// not to have one.
+    ensures all(domain<type>(), |ty|
+              all(domain<address>(), |addr|
+                !old(exists<T<ty>>(addr)) ==> !exists<T<ty>>(addr)));
+  }
+  spec module {
+    /// Apply OnlyCreateCanCreateOffer
+    apply OnlyCreateCanCreateOffer to *<Offered>, * except create;
+  }
+  spec fun create {
+    /// Offer a struct to the account under address `for` by
+    /// placing the Offer under the signer's address
+    aborts_if exists<T<Offered>>(Signer::get_address(account));
+    ensures exists<T<Offered>>(Signer::get_address(account));
+    ensures global<T<Offered>>(Signer::get_address(account)) == T<Offered> { offered: offered, for: for };
+  }
+
+  // Switch documentation context back to module level.
+  spec module {}
+
+  /// ## Removal of Offers
+
+  spec schema OnlyRedeemCanRemoveOffer {
+    /// Only `Self::redeem` can remove an Offer.T.
+    ///
+    /// **Informally:** No other function except for `redeem` can remove an Offer from an account
+    ensures all(domain<type>(), |ty|
+              all(domain<address>(), |addr|
+                old(exists<T<ty>>(addr))
+                  ==> (exists<T<ty>>(addr) && global<T<ty>>(addr) == old(global<T<ty>>(addr)))));
+  }
+  spec module {
+    /// Show that every function except `Self::redeem` can remove an Offer.T from the global store
+    apply OnlyRedeemCanRemoveOffer to *<Offered>, * except redeem;
+  }
+  spec fun redeem {
+    /// Redeems an offer (T) under the account at `offer_address`
```

Comment:
More like a comment which would belong to the function definition itself. We should consider the entire module to be what we are working on, so perhaps move there.